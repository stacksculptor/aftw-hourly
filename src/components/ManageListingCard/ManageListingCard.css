@import '../../marketplace.css';

:root {
  --coverEverything: {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
}

.root {
  /* Layout */
  display: flex;
  flex-direction: column;
  position: relative;

  /* Remove link's hover effect */
  &:hover {
    text-decoration: none;
  }
}

.menuOverlayWrapper {
  /* Positioning */
  @apply --coverEverything;
}

.menuOverlay {
  /* Positioning */
  @apply --coverEverything;

  /* Overlay */
  background-color: var(--matterColorAnti);
  mix-blend-mode: multiply;
  opacity: 0;
  transition: var(--transitionStyleButton);
}

.menuOverlayContent {
  /* Positioning */
  @apply --coverEverything;
  /* Fonts */
  @apply --marketplaceH4FontStyles;
  font-weight: var(--fontWeightSemiBold);
  color: var(--matterColorLight);

  /* Layout */
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 0;
  margin-bottom: 0;
  opacity: 0;

  @media (--viewportMedium) {
    font-weight: var(--fontWeightSemiBold);
    margin-top: 0;
    margin-bottom: 0;
  }
}

.threeToTwoWrapper {
  /* Layout */
  display: block;
  width: 100%;
  position: relative;

  &:hover .menuOverlay {
    opacity: 0.8;
  }

  &:hover .menuOverlayContent {
    opacity: 1.0;
  }
}

.menuOverlayOpen {
  & .menuOverlay {
    opacity: 0.8;
  }

  & .menuOverlayContent {
    opacity: 0.8;
  }
}

/* Firefox doesn't support image aspect ratio inside flexbox */
.aspectWrapper {
  padding-bottom: 66.6667%; /* 3:2 Aspect Ratio */
  background: var(--matterColorNegative); /* Loading BG color */
}

.rootForImage {
  /* Layout - image will take space defined by aspect ratio wrapper */
  @apply --coverEverything;
  width: 100%;
  border-radius: 2px;
}

.menubarWrapper {
  /* Position */
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 42px;
}

.menubarGradient {
  /* Position */
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  /* Gradient */
  background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAqCAYAAACUV/o4AAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAALEwAACxMBAJqcGAAAAVlpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDUuNC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6dGlmZj0iaHR0cDovL25zLmFkb2JlLmNvbS90aWZmLzEuMC8iPgogICAgICAgICA8dGlmZjpPcmllbnRhdGlvbj4xPC90aWZmOk9yaWVudGF0aW9uPgogICAgICA8L3JkZjpEZXNjcmlwdGlvbj4KICAgPC9yZGY6UkRGPgo8L3g6eG1wbWV0YT4KTMInWQAAAGhJREFUKBVjZGBgKANibiAGga8sQAKEf4N4QMAC4vwH4r8gHgiABOAcmMAfEAMGsKrA0DIqAAsvHGFKkwACRRQjzGJQRP2DcUA0LOZAEQ5S9R8mAJIEA6wCyIYSqYVSz4FcxwR1E5gCAD7HLabS/HFUAAAAAElFTkSuQmCC');
  background-repeat: repeat;
  opacity: 0.5;
}

.menubar {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: flex-end;
}

.menuLabel {
  padding: 8px 8px 0 8px;
}

.iconWrapper {
  padding: 2px 10px 8px 10px;
  color: var(--matterColorLight);
  border-radius: 2px 2px 0 0;
}

.listingMenuIsOpen {
  & .iconWrapper {
    background-color: var(--matterColorLight);
    color: var(--matterColor);
  }
}

.menuContent {
  visibility:hidden;
  opacity: 0;
  pointer-events: none;

  position: absolute;
  right: 0;
  z-index: var(--zIndexPopup);

  background-color: var(--matterColorLight);
  border-radius: 2px 0 2px 2px;
}

.menuItem {
  @apply --marketplaceH4FontStyles;
  color: var(--matterColor);
  white-space: nowrap;
  text-align: left;
  min-width: 175px;
  padding: 8px 14px;

  /* Remove default margins from font */
  margin-top: 0;
  margin-bottom: 0;

  &:hover {
    color: var(--failColor);
    text-decoration: none;
  }

  @media (--viewportMedium) {
    margin-top: 0;
    margin-bottom: 0;
  }
}

.menuItemDisabled {
  color: var(--matterColorNegative);

  &:hover {
    color: var(--matterColorNegative);
  }
}

.info {
  /* Layout */
  display: flex;
  flex-direction: row;
  padding: 16px 0 2px 0;
}

.price {
  /* Layout */
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  flex-shrink: 0;
  margin-right: 18px;
}

.priceValue {
  /* Font */
  @apply --marketplaceH3FontStyles;

  /* Remove default margins from font */
  margin-top: 0;
  margin-bottom: 0;

  @media (--viewportMedium) {
    margin-top: 0;
    margin-bottom: 0;
  }
}

.perNight {
  /* Font */
  @apply --marketplaceTinyFontStyles;
  color: var(--matterColor);

  /* Remove default margins from font */
  margin-top: 0;
  margin-bottom: 0;

  @media (--viewportMedium) {
    margin-top: 0;
    margin-bottom: 0;
  }
}


.mainInfo {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
}

.title {
  /* Font */
  @apply --marketplaceH3FontStyles;
  color: var(--matterColor);

  /* Remove default margins from font */
  margin-top: 0;
  margin-bottom: 0;

  @media (--viewportMedium) {
    margin-top: 0;
    margin-bottom: 0;
  }
}

.edit {
  /* TODO: Odd font-size */
  @apply --marketplaceButtonStylesSecondary;
  font-size: 14px;

  /* Reserve space for button */
  flex-shrink: 0;
  width: auto;
  height: 41px;
  padding: 7px 14px 7px 35px;

  /* Add edit icon as a background image */
  background-image: url('data:image/svg+xml;utf8,<svg width="14" height="14" viewBox="0 0 14 14" xmlns="http://www.w3.org/2000/svg"><g stroke="#4A4A4A" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"><path d="M5.307 11.155L1 13l1.846-4.308L10.54 1 13 3.46zM11 5L9 3M5 11L3 9"/></g></svg>');
  background-repeat: no-repeat;
  background-position: 15px center;
}

.loadingOverlayWrapper,
.errorOverlayWrapper,
.closedOverlayWrapper {
  /* Positioning */
  @apply --coverEverything;
}

.loadingOverlay,
.errorOverlay,
.closedOverlay {
  /* Positioning */
  @apply --coverEverything;

  /* Overlay background style */
  background-color: var(--matterColorLight);
  opacity: 0.9;
}

.loadingOverlayContent,
.errorOverlayContent,
.closedOverlayContent {
  @apply --marketplaceH4FontStyles;
  color: var(--matterColor);

  /* Positioning */
  @apply --coverEverything;

  /* Layout */
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: default;

  /* Remove default margins from font */
  margin-top: 0;
  margin-bottom: 0;

  @media (--viewportMedium) {
    margin-top: 0;
    margin-bottom: 0;
  }
}

.errorOverlayContent {
  color: var(--failColor);
  justify-content: flex-start;
  margin-top: 90px;
}

.closedMessage {
  max-width: 220px;
  text-align: center;
}

.openListingButton {
  @apply --marketplaceH5FontStyles;
  width: 114px;
  height: 41px;
  padding: 8px 0;
}
