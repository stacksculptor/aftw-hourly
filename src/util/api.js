import { types as sdkTypes, transit } from './sdkLoader';
import Decimal from 'decimal.js';

const apiBaseUrl = () => {
  const port = process.env.REACT_APP_DEV_API_SERVER_PORT;
  const useDevApiServer = process.env.NODE_ENV === 'development' && !!port;

  // In development, the dev API server is running in a different port
  if (useDevApiServer) {
    return `http://localhost:${port}`;
  }

  // Otherwise, use the same domain and port as the frontend
  return '/';
};

const typeHandlers = [
  {
    type: sdkTypes.BigDecimal,
    customType: Decimal,
    writer: v => new sdkTypes.BigDecimal(v.toString()),
    reader: v => new Decimal(v.value),
  },
];

const serialize = data => {
  return transit.write(data, { typeHandlers });
};

const deserialize = str => {
  return transit.read(str, { typeHandlers });
};

const post = (path, body) => {
  const url = `${apiBaseUrl()}${path}`;
  const options = {
    method: 'POST',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/transit+json',
    },
    body: serialize(body),
  };
  return window
    .fetch(url, options)
    .then(res => {
      if (res.status >= 400) {
        const e = new Error('Local API request failed');
        e.apiResponse = res;
        throw e;
      }
      return res;
    })
    .then(res => {
      const contentTypeHeader = res.headers.get('Content-Type');
      const contentType = contentTypeHeader ? contentTypeHeader.split(';')[0] : null;
      if (contentType === 'application/transit+json') {
        return res.text().then(deserialize);
      } else if (contentType === 'application/json') {
        return res.json();
      }
      return res.text();
    });
};

export const transactionLineItems = body => {
  return post('/api/transaction-line-items', body);
};

export const initiatePrivileged = body => {
  return post('/api/initiate-privileged', body);
};
