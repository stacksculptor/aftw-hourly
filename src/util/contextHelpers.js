import React, { PropTypes, Component as ReactComponent } from 'react'; // eslint-disable-line react/no-deprecated
import { throttle } from 'lodash';

/**
 * A higher order component (HOC) to take the togglePageClassNames function from
 * the context that the Page component has provided.
 */
export const withTogglePageClassNames = Component => {
  const WithTogglePageClassNamesComponent = (props, context) => (
    <Component togglePageClassNames={context.togglePageClassNames} {...props} />
  );

  WithTogglePageClassNamesComponent.displayName = `withTogglePageClassNames(${Component.displayName || Component.name})`;

  const { func } = PropTypes;

  WithTogglePageClassNamesComponent.contextTypes = {
    togglePageClassNames: func.isRequired,
  };

  return WithTogglePageClassNamesComponent;
};

/**
 * A higher order component (HOC) that provides the current viewport
 * dimensions to the wrapped component as a `viewport` prop that has
 * the shape `{ width: 600, height: 400}`.
 */
export const withViewport = Component => {
  // The resize event is flooded when the browser is resized. We'll
  // use a small timeout to throttle changing the viewport since it
  // will trigger rerendering.
  const WAIT_MS = 100;

  class WithViewportComponent extends ReactComponent {
    constructor(props) {
      super(props);
      this.state = { width: 0, height: 0 };
      this.handleWindowResize = this.handleWindowResize.bind(this);
      this.setViewport = throttle(this.setViewport.bind(this), WAIT_MS);
    }
    componentDidMount() {
      this.setViewport();
      window.addEventListener('resize', this.handleWindowResize);
    }
    componentWillUnmount() {
      window.removeEventListener('resize', this.handleWindowResize);
    }
    handleWindowResize() {
      this.setViewport();
    }
    setViewport() {
      this.setState({
        width: window.innerWidth,
        height: window.innerHeight,
      });
    }
    render() {
      const viewport = this.state;
      const props = { ...this.props, viewport };
      return <Component {...props} />;
    }
  }

  WithViewportComponent.displayName = `withViewport(${Component.displayName || Component.name})`;

  return WithViewportComponent;
};
