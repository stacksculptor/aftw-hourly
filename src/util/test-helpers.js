import React from 'react';
import { IntlProvider } from 'react-intl';
import { BrowserRouter } from 'react-router';
import { Provider } from 'react-redux';
import configureStore from '../store';

export const TestProvider = props => {
  const store = configureStore();
  return (
    <IntlProvider locale="en">
      <BrowserRouter>
        <Provider store={store}>
          {props.children}
        </Provider>
      </BrowserRouter>
    </IntlProvider>
  );
};
