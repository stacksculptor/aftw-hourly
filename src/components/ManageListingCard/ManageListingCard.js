import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { withRouter } from 'react-router-dom';
import { FormattedMessage, intlShape, injectIntl } from 'react-intl';
import classNames from 'classnames';
import routeConfiguration from '../../routeConfiguration';
import { LISTING_STATE_PENDING_APPROVAL, LISTING_STATE_CLOSED, propTypes } from '../../util/types';
import { formatMoney } from '../../util/currency';
import { ensureOwnListing } from '../../util/data';
import { LISTING_PAGE_PENDING_APPROVAL_VARIANT, createSlug } from '../../util/urlHelpers';
import { createResourceLocatorString } from '../../util/routes';
import {
  InlineTextButton,
  Menu,
  MenuLabel,
  MenuContent,
  MenuItem,
  IconSpinner,
  ResponsiveImage,
} from '../../components';
import config from '../../config';

import MenuIcon from './MenuIcon';
import css from './ManageListingCard.css';

// Menu content needs the same padding
const MENU_CONTENT_OFFSET = -12;
const MAX_LENGTH_FOR_WORDS_IN_TITLE = 7;

const priceData = (price, intl) => {
  if (price && price.currency === config.currency) {
    const formattedPrice = formatMoney(intl, price);
    return { formattedPrice, priceTitle: formattedPrice };
  } else if (price) {
    return {
      formattedPrice: intl.formatMessage(
        { id: 'ManageListingCard.unsupportedPrice' },
        { currency: price.currency }
      ),
      priceTitle: intl.formatMessage(
        { id: 'ManageListingCard.unsupportedPriceTitle' },
        { currency: price.currency }
      ),
    };
  }
  return {};
};

const createEditListingURL = (routes, listing) => {
  const id = listing.id.uuid;
  const slug = createSlug(listing.attributes.title);
  const pathParams = { id, slug, type: 'edit', tab: 'description' };

  return createResourceLocatorString('EditListingPage', routes, pathParams, {});
};

const createListingURL = (routes, listing) => {
  const id = listing.id.uuid;
  const slug = createSlug(listing.attributes.title);
  const isPendingApproval = listing.attributes.state === LISTING_STATE_PENDING_APPROVAL;
  const linkProps = isPendingApproval
    ? {
        name: 'ListingPageVariant',
        params: {
          id,
          slug,
          variant: LISTING_PAGE_PENDING_APPROVAL_VARIANT,
        },
      }
    : {
        name: 'ListingPage',
        params: { id, slug },
      };

  return createResourceLocatorString(linkProps.name, routes, linkProps.params, {});
};

// Cards are not fixed sizes - So, long words in title make flexboxed items to grow too big.
// 1. We split title to an array of words and spaces.
//    "foo bar".split(/([^\s]+)/gi) => ["", "foo", " ", "bar", ""]
// 2. Then we break long words by adding a '<span>' with word-break: 'break-all';
const formatTitle = (title, maxLength) => {
  const nonWhiteSpaceSequence = /([^\s]+)/gi;
  return title.split(nonWhiteSpaceSequence).map((word, index) => {
    return word.length > maxLength ? (
      <span key={index} style={{ wordBreak: 'break-all' }}>
        {word}
      </span>
    ) : (
      word
    );
  });
};

export const ManageListingCardComponent = props => {
  const {
    className,
    rootClassName,
    hasClosingError,
    hasOpeningError,
    history,
    intl,
    isMenuOpen,
    actionsInProgressListingId,
    listing,
    onCloseListing,
    onOpenListing,
    onToggleMenu,
    renderSizes,
  } = props;
  const classes = classNames(rootClassName || css.root, className);
  const currentListing = ensureOwnListing(listing);
  const id = currentListing.id.uuid;
  const { title = '', price, state } = currentListing.attributes;
  const isPendingApproval = state === LISTING_STATE_PENDING_APPROVAL;
  const isClosed = state === LISTING_STATE_CLOSED;
  const firstImage =
    currentListing.images && currentListing.images.length > 0 ? currentListing.images[0] : null;

  const menuItemClasses = classNames(css.menuItem, {
    [css.menuItemDisabled]: !!actionsInProgressListingId,
  });

  const { formattedPrice, priceTitle } = priceData(price, intl);

  /* eslint-disable jsx-a11y/no-static-element-interactions */
  const closedOverlay = !isClosed ? null : (
    <div
      className={css.closedOverlayWrapper}
      onClick={event => {
        event.preventDefault();
        event.stopPropagation();
      }}
    >
      <div className={css.closedOverlay} />
      <div className={css.closedOverlayContent}>
        <div className={css.closedMessage}>
          <FormattedMessage id="ManageListingCard.closedListing" />
        </div>
        <button
          className={css.openListingButton}
          disabled={!!actionsInProgressListingId}
          onClick={event => {
            event.preventDefault();
            event.stopPropagation();
            if (!actionsInProgressListingId) {
              onOpenListing(currentListing.id);
            }
          }}
        >
          <FormattedMessage id="ManageListingCard.openListing" />
        </button>
      </div>
    </div>
  );

  const errorOverlay =
    hasOpeningError || hasClosingError ? (
      <div
        className={css.errorOverlayWrapper}
        onClick={event => {
          event.preventDefault();
          event.stopPropagation();
        }}
      >
        <div className={css.errorOverlay} />
        <div className={css.errorOverlayContent}>
          <div className={css.closedMessage}>
            <FormattedMessage id="ManageListingCard.actionFailed" />
          </div>
        </div>
      </div>
    ) : null;

  const pendingApprovalOverlay = isPendingApproval ? (
    <div
      className={css.pendingApprovalOverlayWrapper}
      onClick={event => {
        event.preventDefault();
        event.stopPropagation();
      }}
    >
      <div className={css.pendingApprovalOverlay} />
      <div className={css.pendingApprovalOverlayContent}>
        <div className={css.pendingMessage}>
          <FormattedMessage
            id="ManageListingCard.pendingApproval"
            values={{ listingTitle: title }}
          />
        </div>
      </div>
    </div>
  ) : null;

  const thisInProgress = actionsInProgressListingId && actionsInProgressListingId.uuid === id;
  const loadingOrErrorOverlay = thisInProgress ? (
    <div
      className={css.loadingOverlayWrapper}
      onClick={event => {
        event.preventDefault();
        event.stopPropagation();
      }}
    >
      <div className={css.loadingOverlay} />
      <div className={css.loadingOverlayContent}>
        <IconSpinner />
      </div>
    </div>
  ) : (
    errorOverlay
  );
  /* eslint-enable jsx-a11y/no-static-element-interactions */

  const titleClasses = classNames(css.title, {
    [css.titlePending]: isPendingApproval,
  });

  return (
    <div
      className={classes}
      tabIndex={0}
      onClick={event => {
        event.preventDefault();
        event.stopPropagation();

        // ManageListingCard contains links, buttons and elements that are working with routing.
        // This card doesn't work if <a> or <button> is used to wrap events that are card 'clicks'.
        //
        // NOTE: It might be better to absolute-position those buttons over a card-links.
        // (So, that they have no parent-child relationship - like '<a>bla<a>blaa</a></a>')
        history.push(createListingURL(routeConfiguration(), listing));
      }}
    >
      <div className={css.threeToTwoWrapper}>
        <div className={css.aspectWrapper}>
          <ResponsiveImage
            rootClassName={css.rootForImage}
            alt={title}
            image={firstImage}
            variants={['landscape-crop', 'landscape-crop2x']}
            sizes={renderSizes}
          />
        </div>
        <div className={classNames(css.menuOverlayWrapper, { [css.menuOverlayOpen]: isMenuOpen })}>
          <div className={classNames(css.menuOverlay)} />
          <div className={css.menuOverlayContent}>
            <FormattedMessage id="ManageListingCard.viewListing" />
          </div>
        </div>
        <div className={css.menubarWrapper}>
          <div className={css.menubarGradient} />
          <div className={css.menubar}>
            <Menu
              className={classNames(css.menu, { [css.cardIsOpen]: !isClosed })}
              contentPlacementOffset={MENU_CONTENT_OFFSET}
              contentPosition="left"
              useArrow={false}
              onToggleActive={isOpen => {
                const listingOpen = isOpen ? currentListing : null;
                onToggleMenu(listingOpen);
              }}
              isOpen={isMenuOpen}
            >
              <MenuLabel className={css.menuLabel} isOpenClassName={css.listingMenuIsOpen}>
                <div className={css.iconWrapper}>
                  <MenuIcon className={css.menuIcon} isActive={isMenuOpen} />
                </div>
              </MenuLabel>
              <MenuContent rootClassName={css.menuContent}>
                <MenuItem key="close-listing">
                  <InlineTextButton
                    className={menuItemClasses}
                    onClick={event => {
                      event.preventDefault();
                      event.stopPropagation();
                      if (!actionsInProgressListingId) {
                        onToggleMenu(null);
                        onCloseListing(currentListing.id);
                      }
                    }}
                  >
                    <FormattedMessage id="ManageListingCard.closeListing" />
                  </InlineTextButton>
                </MenuItem>
              </MenuContent>
            </Menu>
          </div>
        </div>
        {closedOverlay}
        {pendingApprovalOverlay}
        {loadingOrErrorOverlay}
      </div>
      <div className={css.info}>
        <div className={css.price}>
          <div className={css.priceValue} title={priceTitle}>
            {formattedPrice}
          </div>
          <div className={css.perUnit}>
            <FormattedMessage id="ManageListingCard.perUnit" />
          </div>
        </div>
        <div className={css.mainInfo}>
          <div className={titleClasses}>{formatTitle(title, MAX_LENGTH_FOR_WORDS_IN_TITLE)}</div>
        </div>
        <button
          className={css.edit}
          onClick={event => {
            event.preventDefault();
            event.stopPropagation();
            history.push(createEditListingURL(routeConfiguration(), listing));
          }}
        >
          <FormattedMessage id="ManageListingCard.edit" />
        </button>
      </div>
    </div>
  );
};

ManageListingCardComponent.defaultProps = {
  className: null,
  rootClassName: null,
  actionsInProgressListingId: null,
  renderSizes: null,
};

const { bool, func, shape, string } = PropTypes;

ManageListingCardComponent.propTypes = {
  className: string,
  rootClassName: string,
  hasClosingError: bool.isRequired,
  hasOpeningError: bool.isRequired,
  intl: intlShape.isRequired,
  listing: propTypes.ownListing.isRequired,
  isMenuOpen: bool.isRequired,
  actionsInProgressListingId: shape({ uuid: string.isRequired }),
  onCloseListing: func.isRequired,
  onOpenListing: func.isRequired,
  onToggleMenu: func.isRequired,

  // Responsive image sizes hint
  renderSizes: string,

  // from withRouter
  history: shape({
    push: func.isRequired,
  }).isRequired,
};

export default compose(withRouter, injectIntl)(ManageListingCardComponent);
