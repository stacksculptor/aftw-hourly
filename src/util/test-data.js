import Decimal from 'decimal.js';
import { types } from './sdkLoader';
import { nightsBetween } from '../util/dates';
import * as propTypes from '../util/propTypes';

const { UUID, LatLng, Money } = types;

// Create a booking that conforms to the util/propTypes booking schema
export const createBooking = (id, attributes = {}) => ({
  id: new UUID(id),
  type: 'booking',
  attributes: {
    start: new Date(Date.UTC(2017, 5, 10)),
    end: new Date(Date.UTC(2017, 5, 10)),
    ...attributes,
  },
});

// Create a user that conforms to the util/propTypes user schema
export const createUser = id => ({
  id: new UUID(id),
  type: 'user',
  attributes: {
    banned: false,
    profile: {
      displayName: `${id} display name`,
      abbreviatedName: `${id} abbreviated name`,
    },
  },
});

// Create a user that conforms to the util/propTypes currentUser schema
export const createCurrentUser = id => ({
  id: new UUID(id),
  type: 'current-user',
  attributes: {
    banned: false,
    email: `${id}@example.com`,
    emailVerified: true,
    profile: {
      firstName: `${id} first name`,
      lastName: `${id} last name`,
      displayName: `${id} display name`,
      abbreviatedName: `${id} abbreviated name`,
    },
    stripeConnected: true,
  },
});

// Create a user that conforms to the util/propTypes user schema
export const createImage = id => ({
  id: new UUID(id),
  type: 'image',
  attributes: {
    sizes: [
      {
        name: 'square',
        height: 408,
        width: 408,
        url: 'https://placehold.it/408x408',
      },
      {
        name: 'square2x',
        height: 816,
        width: 816,
        url: 'https://placehold.it/816x816',
      },
    ],
  },
});

// Create a user that conforms to the util/propTypes listing schema
export const createListing = (id, attributes = {}, includes = {}) => ({
  id: new UUID(id),
  type: 'listing',
  attributes: {
    title: `${id} title`,
    description: `${id} description`,
    address: `${id} address`,
    geolocation: new LatLng(40, 60),
    closed: false,
    deleted: false,
    price: new Money(5500, 'USD'),
    ...attributes,
  },
  ...includes,
});

export const createTransaction = options => {
  const {
    id,
    lastTransition = propTypes.TX_TRANSITION_PREAUTHORIZE,
    total = new Money(1000, 'USD'),
    commission = new Money(100, 'USD'),
    booking = null,
    listing = null,
    customer = null,
    provider = null,
    lastTransitionedAt = new Date(Date.UTC(2017, 5, 1)),
  } = options;
  const nightCount = booking ? nightsBetween(booking.attributes.start, booking.attributes.end) : 1;
  return {
    id: new UUID(id),
    type: 'transaction',
    attributes: {
      createdAt: new Date(Date.UTC(2017, 4, 1)),
      lastTransitionedAt,
      lastTransition,
      payinTotal: total,
      payoutTotal: new Money(total.amount - commission.amount, total.currency),
      lineItems: [
        {
          code: 'line-item/night',
          quantity: new Decimal(nightCount),
          unitPrice: new Money(total.amount / nightCount, total.currency),
          lineTotal: total,
          reversal: false,
        },
        {
          code: 'line-item/provider-commission',
          unitPrice: new Money(commission.amount * -1, commission.currency),
          lineTotal: new Money(commission.amount * -1, commission.currency),
          reversal: false,
        },
      ],
    },
    booking,
    listing,
    customer,
    provider,
  };
};

// Default config for currency formatting in tests and examples.
export const currencyConfig = {
  style: 'currency',
  currency: 'USD',
  currencyDisplay: 'symbol',
  useGrouping: true,
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
};

const pad = num => {
  return num >= 0 && num < 10 ? `0${num}` : `${num}`;
};

// Create fake Internalization object to help with shallow rendering.
export const fakeIntl = {
  formatDate: d => `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}`,
  formatHTMLMessage: d => d,
  formatMessage: msg => msg.id,
  formatNumber: d => d,
  formatPlural: d => d,
  formatRelative: d => d,
  formatTime: d => `${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}`,
  now: () => Date.UTC(2017, 10, 23, 12, 59),
};

const noop = () => null;

export const fakeFormProps = {
  anyTouched: false,
  asyncValidating: false,
  dirty: false,
  form: 'fakeTestForm',
  invalid: false,
  pristine: true,
  clearSubmit: noop,
  touch: noop,
  untouch: noop,
  submit: noop,
  reset: noop,
  initialize: noop,
  handleSubmit: noop,
  destroy: noop,
  clearAsyncError: noop,
  change: noop,
  blur: noop,
  autofill: noop,
  asyncValidate: noop,
  valid: true,
  submitSucceeded: false,
  submitFailed: false,
  submitting: false,
  pure: true,
  initialized: true,
};
