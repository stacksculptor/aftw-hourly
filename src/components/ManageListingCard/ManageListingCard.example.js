/* eslint-disable no-console */
import React from 'react';
import ManageListingCard from './ManageListingCard';
import { createOwnListing, fakeIntl } from '../../util/test-data';

const listing = { ...createOwnListing('listing1') };

const ManageListingCardWrapper = props => (
  <div style={{ maxWidth: '400px' }}>
    <ManageListingCard {...props} />
  </div>
);

export const ManageListingCardWrapped = {
  component: ManageListingCardWrapper,
  props: {
    intl: fakeIntl,
    listing,
  },
};
