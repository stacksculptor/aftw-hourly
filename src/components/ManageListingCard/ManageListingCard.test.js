import React from 'react';
import { renderShallow } from '../../util/test-helpers';
import { createUser, createOwnListing, fakeIntl } from '../../util/test-data';
import { ManageListingCardComponent } from './ManageListingCard';

const noop = () => null;

describe('ManageListingCard', () => {
  it('matches snapshot', () => {
    const listing = { ...createOwnListing('listing1') };
    const tree = renderShallow(
      <ManageListingCardComponent
        flattenedRoutes={[]}
        history={{ push: noop }}
        listing={listing}
        intl={fakeIntl}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
