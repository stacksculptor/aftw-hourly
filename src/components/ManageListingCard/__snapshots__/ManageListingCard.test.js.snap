// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageListingCard matches snapshot 1`] = `
<div
  className=""
  onClick={[Function]}
  tabIndex={0}
>
  <div>
    <div>
      <ResponsiveImage
        alt="listing1 title"
        className={null}
        image={null}
        noImageMessage={null}
        rootClassName={null}
        sizes={null}
        variants={
          Array [
            "landscape-crop",
            "landscape-crop2x",
          ]
        }
      />
    </div>
    <div
      className=""
    >
      <div
        className=""
      />
      <div>
        <FormattedMessage
          id="ManageListingCard.viewListing"
          values={Object {}}
        />
      </div>
    </div>
    <div>
      <div />
      <div>
        <Menu
          className="undefined"
          contentPlacementOffset={-12}
          contentPosition="left"
          isOpen={false}
          onToggleActive={[Function]}
          rootClassName=""
          useArrow={false}
        >
          <MenuLabel
            className={null}
            isOpen={false}
            isOpenClassName={null}
            onToggleActive={null}
            rootClassName=""
          >
            <div>
              <MenuIcon
                className={null}
                isActive={false}
              />
            </div>
          </MenuLabel>
          <MenuContent
            arrowPosition={null}
            className={null}
            contentClassName={null}
            contentRef={null}
            isOpen={false}
            rootClassName=""
            style={null}
          >
            <MenuItem
              className={null}
              key="close-listing"
              rootClassName=""
            >
              <InlineTextButton
                className=""
                onClick={[Function]}
              >
                <FormattedMessage
                  id="ManageListingCard.closeListing"
                  values={Object {}}
                />
              </InlineTextButton>
            </MenuItem>
          </MenuContent>
        </Menu>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        title={55}
      >
        55
      </div>
      <div>
        <FormattedMessage
          id="ManageListingCard.perUnit"
          values={Object {}}
        />
      </div>
    </div>
    <div>
      <div
        className=""
      >
        <span
          key="1"
          style={
            Object {
              "wordBreak": "break-all",
            }
          }
        >
          listing1
        </span>
         
        title
      </div>
    </div>
    <button
      onClick={[Function]}
    >
      <FormattedMessage
        id="ManageListingCard.edit"
        values={Object {}}
      />
    </button>
  </div>
</div>
`;
